django/db/models/base.py

```diff
@@ -630,6 +630,18 @@ def save(self, force_insert=False, force_update=False, using=None,
         that the "save" must be an SQL insert or update (or equivalent for
         non-SQL backends), respectively. Normally, they should not be set.
         """
+        for field in self._meta.concrete_fields:
+            if field.is_relation:
+                try:
+                    obj = getattr(self, field.name, None)
```

Comment:
I share your concern about database queries and indeed that is a cause of some of the tests failures, however, I think the assumption about unsaved fields haven't `None` values doesn't hold in cases like `UUIDField` (see [ticket #24691](https://code.djangoproject.com/ticket/24691)). Anyway, I pushed that change and a new test.

I wasn't making very quick progress on this ticket so I pushed my WIP in hopes someone familiar with the internal model APIs might be able to do better.